File: console/src/services/api/webhook_subscription.ts
import { api } from './client'

export interface CustomEventFilters {
  goal_types?: string[]
  event_names?: string[]
}

export interface WebhookSubscriptionSettings {
  event_types: string[]
  custom_event_filters?: CustomEventFilters
}

export interface WebhookSubscription {
  id: string
  name: string
  url: string
  secret: string
  settings: WebhookSubscriptionSettings
  // Flattened from settings by backend MarshalJSON
  event_types?: string[]
  custom_event_filters?: CustomEventFilters
  enabled: boolean
  last_delivery_at?: string
  created_at: string
  updated_at: string
}

export interface WebhookDelivery {
  id: string
  subscription_id: string
  event_type: string
  payload: Record<string, unknown>
  status: 'pending' | 'delivered' | 'failed'
  attempts: number
  max_attempts: number
  next_attempt_at: string
  last_attempt_at?: string
  delivered_at?: string
  last_response_status?: number
  last_response_body?: string
  last_error?: string
  created_at: string
}

export interface CreateWebhookSubscriptionRequest {
  workspace_id: string
  name: string
  url: string
  event_types: string[]
  custom_event_filters?: CustomEventFilters
}

export interface UpdateWebhookSubscriptionRequest {
  workspace_id: string
  id: string
  name: string
  url: string
  event_types: string[]
  custom_event_filters?: CustomEventFilters
  enabled: boolean
}

export interface ToggleWebhookSubscriptionRequest {
  workspace_id: string
  id: string
  enabled: boolean
}

export interface TestWebhookResponse {
  success: boolean
  status_code: number
  response_body: string
  error?: string
}

export interface GetDeliveriesResponse {
  deliveries: WebhookDelivery[]
  total: number
  limit: number
  offset: number
}

export const webhookSubscriptionApi = {
  create: async (
    params: CreateWebhookSubscriptionRequest
  ): Promise<{ subscription: WebhookSubscription }> => {
    return api.post('/api/webhookSubscriptions.create', params)
  },

  list: async (workspaceId: string): Promise<{ subscriptions: WebhookSubscription[] }> => {
    const searchParams = new URLSearchParams()
    searchParams.append('workspace_id', workspaceId)
    return api.get<{ subscriptions: WebhookSubscription[] }>(
      `/api/webhookSubscriptions.list?${searchParams.toString()}`
    )
  },

  get: async (
    workspaceId: string,
    id: string
  ): Promise<{ subscription: WebhookSubscription }> => {
    const searchParams = new URLSearchParams()
    searchParams.append('workspace_id', workspaceId)
    searchParams.append('id', id)
    return api.get<{ subscription: WebhookSubscription }>(
      `/api/webhookSubscriptions.get?${searchParams.toString()}`
    )
  },

  update: async (
    params: UpdateWebhookSubscriptionRequest
  ): Promise<{ subscription: WebhookSubscription }> => {
    return api.post('/api/webhookSubscriptions.update', params)
  },

  delete: async (workspaceId: string, id: string): Promise<{ success: boolean }> => {
    return api.post('/api/webhookSubscriptions.delete', {
      workspace_id: workspaceId,
      id
    })
  },

  toggle: async (
    params: ToggleWebhookSubscriptionRequest
  ): Promise<{ subscription: WebhookSubscription }> => {
    return api.post('/api/webhookSubscriptions.toggle', params)
  },

  regenerateSecret: async (
    workspaceId: string,
    id: string
  ): Promise<{ subscription: WebhookSubscription }> => {
    return api.post('/api/webhookSubscriptions.regenerateSecret', {
      workspace_id: workspaceId,
      id
    })
  },

  getDeliveries: async (
    workspaceId: string,
    subscriptionId?: string,
    limit?: number,
    offset?: number
  ): Promise<GetDeliveriesResponse> => {
    const searchParams = new URLSearchParams()
    searchParams.append('workspace_id', workspaceId)
    if (subscriptionId) searchParams.append('subscription_id', subscriptionId)
    if (limit !== undefined) searchParams.append('limit', limit.toString())
    if (offset !== undefined) searchParams.append('offset', offset.toString())
    return api.get<GetDeliveriesResponse>(
      `/api/webhookSubscriptions.deliveries?${searchParams.toString()}`
    )
  },

  test: async (
    workspaceId: string,
    id: string,
    eventType: string
  ): Promise<TestWebhookResponse> => {
    return api.post('/api/webhookSubscriptions.test', {
      workspace_id: workspaceId,
      id,
      event_type: eventType
    })
  },

  getEventTypes: async (): Promise<{ event_types: string[] }> => {
    return api.get<{ event_types: string[] }>('/api/webhookSubscriptions.eventTypes')
  }
}

File: internal/domain/webhook_subscription.go
package domain

//go:generate mockgen -destination mocks/mock_webhook_subscription_repository.go -package mocks github.com/Notifuse/notifuse/internal/domain WebhookSubscriptionRepository
//go:generate mockgen -destination mocks/mock_webhook_delivery_repository.go -package mocks github.com/Notifuse/notifuse/internal/domain WebhookDeliveryRepository

import (
	"context"
	"encoding/json"
	"time"
)

// WebhookSubscriptionSettings contains event subscription configuration
type WebhookSubscriptionSettings struct {
	EventTypes         []string            `json:"event_types"`
	CustomEventFilters *CustomEventFilters `json:"custom_event_filters,omitempty"`
}

// WebhookSubscription represents an outgoing webhook subscription configuration
type WebhookSubscription struct {
	ID             string                      `json:"id"`
	Name           string                      `json:"name"`
	URL            string                      `json:"url"`
	Secret         string                      `json:"secret"`
	Settings       WebhookSubscriptionSettings `json:"settings"`
	Enabled        bool                        `json:"enabled"`
	LastDeliveryAt *time.Time                  `json:"last_delivery_at,omitempty"`
	CreatedAt      time.Time                   `json:"created_at"`
	UpdatedAt      time.Time                   `json:"updated_at"`
}

// MarshalJSON implements custom JSON marshaling to flatten settings into top-level fields
// This maintains backward-compatible API responses while using nested internal structure
func (w WebhookSubscription) MarshalJSON() ([]byte, error) {
	type Alias WebhookSubscription
	return json.Marshal(&struct {
		Alias
		EventTypes         []string            `json:"event_types"`
		CustomEventFilters *CustomEventFilters `json:"custom_event_filters,omitempty"`
	}{
		Alias:              Alias(w),
		EventTypes:         w.Settings.EventTypes,
		CustomEventFilters: w.Settings.CustomEventFilters,
	})
}

// CustomEventFilters defines filters for custom event subscriptions
type CustomEventFilters struct {
	GoalTypes  []string `json:"goal_types,omitempty"`  // Filter by goal_type enum
	EventNames []string `json:"event_names,omitempty"` // Filter by event_name
}

// WebhookDelivery represents a pending or completed webhook delivery
type WebhookDelivery struct {
	ID                 string                 `json:"id"`
	SubscriptionID     string                 `json:"subscription_id"`
	EventType          string                 `json:"event_type"`
	Payload            map[string]interface{} `json:"payload"`
	Status             string                 `json:"status"` // pending, delivering, delivered, failed
	Attempts           int                    `json:"attempts"`
	MaxAttempts        int                    `json:"max_attempts"`
	NextAttemptAt      time.Time              `json:"next_attempt_at"`
	LastAttemptAt      *time.Time             `json:"last_attempt_at,omitempty"`
	DeliveredAt        *time.Time             `json:"delivered_at,omitempty"`
	LastResponseStatus *int                   `json:"last_response_status,omitempty"`
	LastResponseBody   *string                `json:"last_response_body,omitempty"`
	LastError          *string                `json:"last_error,omitempty"`
	CreatedAt          time.Time              `json:"created_at"`
}

// WebhookDeliveryStatus constants
const (
	WebhookDeliveryStatusPending    = "pending"
	WebhookDeliveryStatusDelivering = "delivering"
	WebhookDeliveryStatusDelivered  = "delivered"
	WebhookDeliveryStatusFailed     = "failed"
)

// Available webhook event types
var WebhookEventTypes = []string{
	// Contact events
	"contact.created",
	"contact.updated",
	"contact.deleted",
	// List events
	"list.subscribed",
	"list.unsubscribed",
	"list.confirmed",
	"list.resubscribed",
	"list.bounced",
	"list.complained",
	"list.pending",
	"list.removed",
	// Segment events
	"segment.joined",
	"segment.left",
	// Email events
	"email.sent",
	"email.delivered",
	"email.opened",
	"email.clicked",
	"email.bounced",
	"email.complained",
	"email.unsubscribed",
	// Custom events (with optional filtering)
	"custom_event.created",
	"custom_event.updated",
	"custom_event.deleted",
}

// WebhookSubscriptionRepository defines the interface for webhook subscription data access
type WebhookSubscriptionRepository interface {
	Create(ctx context.Context, workspaceID string, sub *WebhookSubscription) error
	GetByID(ctx context.Context, workspaceID, id string) (*WebhookSubscription, error)
	List(ctx context.Context, workspaceID string) ([]*WebhookSubscription, error)
	Update(ctx context.Context, workspaceID string, sub *WebhookSubscription) error
	Delete(ctx context.Context, workspaceID, id string) error
	UpdateLastDeliveryAt(ctx context.Context, workspaceID, id string, deliveredAt time.Time) error
}

// WebhookDeliveryRepository defines the interface for webhook delivery data access
type WebhookDeliveryRepository interface {
	GetPendingForWorkspace(ctx context.Context, workspaceID string, limit int) ([]*WebhookDelivery, error)
	ListAll(ctx context.Context, workspaceID string, subscriptionID *string, limit, offset int) ([]*WebhookDelivery, int, error)
	UpdateStatus(ctx context.Context, workspaceID, id string, status string, attempts int, responseStatus *int, responseBody, lastError *string) error
	MarkDelivered(ctx context.Context, workspaceID, id string, responseStatus int, responseBody string) error
	ScheduleRetry(ctx context.Context, workspaceID, id string, nextAttempt time.Time, attempts int, responseStatus *int, responseBody, lastError *string) error
	MarkFailed(ctx context.Context, workspaceID, id string, attempts int, lastError string, responseStatus *int, responseBody *string) error
	Create(ctx context.Context, workspaceID string, delivery *WebhookDelivery) error
	CleanupOldDeliveries(ctx context.Context, workspaceID string, retentionDays int) (int64, error)
}

// WebhookDeliveryWithSubscription contains a delivery with its associated subscription
type WebhookDeliveryWithSubscription struct {
	Delivery     *WebhookDelivery
	Subscription *WebhookSubscription
	WorkspaceID  string
}
